idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyInDestructuringAssignmentIntention.kt

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2010-2017 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jetbrains.kotlin.idea.intentions
+
+import com.intellij.codeInsight.intention.LowPriorityAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.util.TextRange
+import org.jetbrains.kotlin.idea.core.setType
+import org.jetbrains.kotlin.psi.KtCodeFragment
+import org.jetbrains.kotlin.psi.KtDestructuringDeclaration
+import org.jetbrains.kotlin.psi.psiUtil.endOffset
+import org.jetbrains.kotlin.psi.psiUtil.startOffset
+import org.jetbrains.kotlin.types.isError
+
+class SpecifyTypeExplicitlyInDestructuringAssignmentIntention : SelfTargetingRangeIntention<KtDestructuringDeclaration>(
+        KtDestructuringDeclaration::class.java, "Specify all types explicitly in destructuring assignment"
+), LowPriorityAction {
+
+    override fun applicabilityRange(element: KtDestructuringDeclaration): TextRange? {
+        if (element.containingFile is KtCodeFragment) return null
+        val entries = element.noTypeReferenceEntries()
+        if (entries.isEmpty()) return null
+        if (entries.any { SpecifyTypeExplicitlyIntention.getTypeForDeclaration(it).isError }) return null
+        return TextRange(element.startOffset, element.initializer?.let { it.startOffset - 1 } ?: element.endOffset)
+    }
+
+    override fun applyTo(element: KtDestructuringDeclaration, editor: Editor?) {
+        val entries = element.noTypeReferenceEntries()
+        if (editor != null && element.initializer != null)
```

Comment:
This behaves differently for a destructuring assignment and a destructuring used in a loop:

```
        val map = mapOf(1 to "two")
        for ((key, value) in map) { 
        }
```

The behavior should be the same in both cases.

Reply:
If `&& element.initializer != null` is deleted, the behavior will be same:
```
val map = mapOf(1 to "two")
for ((key, value)<caret> in map) {
}
```

However, destructuring in lambda will not work correctly:
```
val map = mapOf(1 to "two")
map.forEach { (key, value)<caret> ->
}
```
I think the cause is `typeRef.isValid` returns false:
https://github.com/t-kameyama/kotlin/blob/827c57d6ed7492b5f41fbc4ac87b196deebde7bf/idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyIntention.kt#L159
I also confirmed that `typeRef.text` is `Any`.

Is there a way for `typeRef.isValid` to become `true`?

Reply:
Fixed